network_simulator/flow.py

```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+        """
+        Schedule a FlowSendPacketsEvent immediately, with the list of packets
+        to send.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        flow_send_event = FlowSendPacketsEvent(self.flow,
+                                               self.packets_to_send)
+        main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+
+
+
+class FlowSendPacketsEvent(Event):
+    def __init__(self, flow, packets_to_send):
+        """
+        Send data packet from a flow.
+
+        :ivar Flow flow: flow of this event.
+        :ivar list packets_to_send: list of data packets to send.
+        """
+        self.flow = flow
+        self.packets_to_send = packets_to_send
+
+    def run(self, main_event_loop, statistics):
+        """
+        Check that the size of packets combined is smaller than
+        the window size.
+        :param MainEventLoop main_event_loop: main loop.
+        :param Statistics statistics: the Statistics to update.
+        """
+        if len(set(self.flow.packets_in_transit) \
+            + set(self.flow.packets_to_send)) \
+            > self.flow.window_size_packets:
+                # TODO(team): What do we want to throw here?
+                pass
+
+        for curr_packet in self.packets_to_send:
+            if curr_packet.packet_id not in self.flow.packets_in_transit:
+                self.flow.packets_in_transit.add(curr_packet.packet_id)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
+        packet.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # TODO(sharon): Need to fix the initialization after Link subclasses
+        # have been implemented.
+        for curr_packet in self.packets_to_send:
+            device_link_event = DeviceToLinkEvent(curr_packet, False)
+            main_event_loop.schedule_event_with_delay(device_link_event, 0.0)
+
+        # Go through the queue and remove any previously scheduled
+        # FlowTimeoutPacketEvent of the packets in packets_to_send.
+        new_queue = PriorityQueue()
+        while not self.events.empty():
+            curr_tup = self.events.get()
+            curr_event = self.curr_tup[1]
+            if type(curr_event) is FlowTimeoutPacketEvent \
+               and curr_event.packet in self.packets_to_send:
+                continue
+            new_queue.add(curr_tup)
+
+        # Schedule FlowTimeoutPacketEvent for each packet, which will occur
+        # in FLOW_TIMEOUT_SEC from now.
+        for curr_packet in self.packets_to_send:
+            timeout_event = FlowTimeoutPacketEvent(self.flow, curr_packet)
+            main_event_loop.schedule_event_with_delay(timeout_event,
+                                                      FLOW_TIMEOUT_SEC)
+
```

Comment:
nit: keep two spaces between classes and between functions outside of classes. PyCharm will underline PEP8 issues like this in grey by default (I think the "class FlowTimeoutPacketEvent" would be underlined in this case).